Potential crash under WebPageProxy::executeEditCommand()

pascoej opened the pull request:
#### 99d7f658ab048449ea97edacd7a539fac1bc1996
<pre>
Potential crash under WebPageProxy::executeEditCommand()
<a href="https://rdar.apple.com/145887512">rdar://145887512</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=301321">https://bugs.webkit.org/show_bug.cgi?id=301321</a>

Reviewed by Charlie Wolfe.

Hardening fix to avoid calling protectedConnection() when the NetworkProcess
doesn&apos;t have an established connection yet. Also add hasRunningProcess() check
in the callback version&apos;s completion handler.

Additionally, track pending file paths that need to be allowed in
NetworkProcessProxy (similar to allowedFirstPartiesForCookies) and send them
via NetworkProcessCreationParameters when the NetworkProcess launches. This
ensures file access is properly granted even if NetworkProcess crashes and
restarts.

This fix is somewhat speculative as I couldn&apos;t reproduce the crash in testing.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::initializeNetworkProcess):
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::WebPasteboardProxy::grantAccessToCurrentData):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
(WebKit::NetworkProcessProxy::networkProcessDidTerminate):
(WebKit::NetworkProcessProxy::addPendingAllowedFilePaths):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::executeEditCommand):

Canonical link: <a href="https://commits.webkit.org/305173@main">https://commits.webkit.org/305173@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fce2dc1767b9fc15442eb7d6a129194f4a280bb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136849 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9209 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48136 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144585 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89822 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/34a577b4-551a-4cdf-b193-eced35823181) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9912 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9055 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104643 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76375 "Build was cancelled. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/39cf1fc4-1878-45c4-a678-c95e3b8a9700) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139794 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7248 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122612 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85481 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c7cdb81e-cece-43c9-a996-97b90833435f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6891 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4581 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5173 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116215 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147342 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8892 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41366 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113001 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8910 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7470 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113330 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28953 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6808 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118896 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63073 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8940 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36950 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8661 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72506 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8880 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8732 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
```diff
@@ -93,7 +93,15 @@ static PasteboardDataLifetime determineDataLifetime(std::optional<WebPageProxyId
     auto processIdentifier = process.coreProcessIdentifier();
-    return process.protectedWebsiteDataStore()->protectedNetworkProcess()->sendWithAsyncReply(Messages::NetworkProcess::AllowFilesAccessFromWebProcess(processIdentifier, paths), WTFMove(completionHandler));
+    RefPtr dataStore = process.websiteDataStore();
+    RefPtr networkProcess = dataStore->networkProcess();
```
```suggestion
    RefPtr networkProcess = process.protectedWebsiteDataStore()->networkProcess();
```

Seems like we don't need `dataStore` more than once?

sysrqb commented on Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
```diff
@@ -93,7 +93,15 @@ static PasteboardDataLifetime determineDataLifetime(std::optional<WebPageProxyId
-    return process.protectedWebsiteDataStore()->protectedNetworkProcess()->sendWithAsyncReply(Messages::NetworkProcess::AllowFilesAccessFromWebProcess(processIdentifier, paths), WTFMove(completionHandler));
+    RefPtr dataStore = process.websiteDataStore();
+    RefPtr networkProcess = dataStore->networkProcess();
+    networkProcess->addAllowedFilePaths(process, paths);
```
Do we need this, even if `networkProcess->hasConnection()`? Maybe you're recommending that we continue caching the allowed file paths, even when we have a connection to the network process, in case the process crashes?

pascoej replied:
Yes. This is the same pattern used by `allowedFirstPartiesForCookies`.

sysrqb commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -3572,7 +3572,7 @@ void WebPageProxy::executeEditCommand(const String& commandName, const String& a
     auto completionHandler = [weakThis = WeakPtr { *this }, callbackFunction = WTFMove(callbackFunction), commandName, argument, targetFrameID] () mutable {
         RefPtr protectedThis = weakThis.get();
-        if (!protectedThis)
+        if (!protectedThis || !protectedThis->hasRunningProcess())
```
Do you know what the failure behavior is here? Will the rest of the code fail gracefully? Or do we need to queue the edit command until we have a network process connection?

pascoej replied:
If we don't have one, the path will be sent in creation parameters when NP starts.

pascoej committed:
Potential crash under WebPageProxy::executeEditCommand()
rdar://145887512
https://bugs.webkit.org/show_bug.cgi?id=301321

Reviewed by Charlie Wolfe.

Hardening fix to avoid calling protectedConnection() when the NetworkProcess
doesn't have an established connection yet. Also add hasRunningProcess() check
in the callback version's completion handler.

Additionally, track pending file paths that need to be allowed in
NetworkProcessProxy (similar to allowedFirstPartiesForCookies) and send them
via NetworkProcessCreationParameters when the NetworkProcess launches. This
ensures file access is properly granted even if NetworkProcess crashes and
restarts.

This fix is somewhat speculative as I couldn't reproduce the crash in testing.

* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::initializeNetworkProcess):
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.h:
* Source/WebKit/NetworkProcess/NetworkProcessCreationParameters.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::WebPasteboardProxy::grantAccessToCurrentData):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::sendCreationParametersToNewProcess):
(WebKit::NetworkProcessProxy::networkProcessDidTerminate):
(WebKit::NetworkProcessProxy::addPendingAllowedFilePaths):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::executeEditCommand):

Canonical link: https://commits.webkit.org/305173@main
